src/Workspaces/Core/Portable/SemanticModelReuse/AbstractSemanticModelReuseLanguageService.cs

```diff
@@ -106,26 +107,51 @@ protected SyntaxNode GetPreviousBodyNode(SyntaxNode previousRoot, SyntaxNode cur
         }
         else
         {
-            using var pooledCurrentMembers = this.SyntaxFacts.GetMethodLevelMembers(currentRoot);
-            var currentMembers = pooledCurrentMembers.Object;
+            // Walk up the ancestor nodes of currentBodyNode, finding child indexes up to the root.
+            using var _ = ArrayBuilder<int>.GetInstance(out var indexPath);
+            GetNodeChildIndexPathToRootReversed(currentBodyNode, indexPath);
 
-            var index = currentMembers.IndexOf(currentBodyNode);
-            if (index < 0)
+            // Then use those indexes to walk back down the previous tree to find the equivalent node.
+            var previousNode = previousRoot;
+            for (var i = indexPath.Count - 1; i >= 0; i--)
             {
-                Debug.Fail($"Unhandled member type in {nameof(GetPreviousBodyNode)}");
-                return null;
+                var childIndex = indexPath[i];
+                var children = previousNode.ChildNodesAndTokens();
+
+                if (children.Count <= childIndex)
+                {
+                    Debug.Fail("Member count shouldn't have changed as there were no top level edits.");
+                    return null;
+                }
+
```

Comment:
Blind AsNode also feels unsafe.

Reply:
I can check it, but it seems safe to me as GetNodeChildIndexPathToRootReversed only adds the index if the AsNode check passes. Or is there a case I'm not considering?

Reply:
oh, duh, I see. Yeah, I'll add a check just to make sure.